### PR TITLE
[SPIR-V] Add SPIR-V structurizer

### PR DESCRIPTION
**Pull request by Keenuts**
This commit adds an initial SPIR-V structurizer.
It leverages the previously merged passes, and the convergence region analysis to determine the correct merge and continue blocks for SPIR-V.

The first part does a branch cleanup (simplifying switches, and legalizing them), then merge instructions are added to cycles, convergent and later divergent blocks.
Then comes the important part: splitting critical edges, and making sure the divergent construct boundaries don't cross.

- we split blocks with multiple headers into 2 blocks.
- we split blocks that are a merge blocks for 2 or more constructs: SPIR-V spec disallow a merge block to be shared by 2 loop/switch/condition construct.
- we split merge & continue blocks: SPIR-V spec disallow a basic block to be both a continue block, and a merge block.
- we remove superfluous headers: when a header doesn't bring more info than the parent on the divergence state, it must be removed.

This PR leverages the merged SPIR-V simulator for testing, as long as spirv-val. For now, most DXC structurization tests are passing. The unsupported ones are either caused by unsupported features like switches on boolean types, or switches in region exits, because the MergeExit pass doesn't support those yet (there is a FIXME).

This PR is quite large, and the addition not trivial, so I tried to keep it simple. E.G: as soon as the CFG changes, I recompute the dominator trees and other structures instead of updating them.

### CODE REVIEW

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (s-perron) [inline llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:783]: I believe infinite loops are not are undefined behaviour. We should be able to done something if we need to. I would not raise an error. It is possible the infinite loop is a by-product of optimizations, and it will never get executed. In general, you cannot check this at compile time.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (s-perron) [inline llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:843]:
```suggestion
  // When a block has multiple OpSelectionMerge/OpLoopMerge instructions, sorts them
  // to put the "largest" first. A merge instruction is defined as
```

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (s-perron) [inline llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:1164]:
```suggestion
  // Makes sure every case target in |F| is unique. If 2 cases branch to the
  // same basic block, one of the targets is updated so it jumps to a new basic
  // block ending with a single unconditional branch to the original target.
```

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (s-perron) [inline llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:1165]: Depending on how you do this, you could get something that cannot be represented in SPIR-V:

```
switch (i) {
  case 0:
  case 1:
  case 2:
    DoSomething();
    break;
  default:
    break;
}
```

This is pretty common. I'm wondering if this funciton would turn it into:

```
switch (i) {
  case 0:
    goto case 2;
  case 1:
    goto case 2;
  case 2:
    DoSomething();
    break;
  default:
    break;
}
```

This is a problem for SPIR-V because case 0 no longer fall through to the next case. A case in spir-v can only branch to another case if it is the next one. We need to make sure that case 0 branches to case 1, and case 1 branches to case 2.

Comment (s-perron) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:1165]: We can deal with this in a follow up PR if something needs to be done.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (s-perron) [inline llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:1327]: I though you could repeat the same target bb multiple times in the same OpSwitch. See the code that glslang generates:

https://shader-playground.timjones.io/ba9ec8e0afb39a9fff42a1d0c0d6baa5

Comment (s-perron) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:1327]: If you need it for the algorithm, then that is something different.

**Review by s-perron — APPROVED** [review]
This looks good. No major issue. There are a few typos, and a couple smaller issues that we can follow up on in another PR.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (s-perron) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:1327]: We can follow up in another PR.

Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:1327]: Yes, this constraint is only valid for if conditions. I'll update the comment to explain I also need this to avoid switches when not requires, as switch have a hard requirements for en OpSelectionMerge while if can avoid it in some cases.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:1165]: Ahh, that's not great.
But I think we should get rid of switched entirely anyway. When I was discussing with Alan, he explain that convergence is now correctly defined for if/else, but switches still have some level of under-defined convergence (fallthrough is one case). So if we want to use maximal reconvergence, and have a 100% defined behavior, I'll probably need to convert those to if/else trees.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:500]: Changed this to check for `assert(L->getLoopLatch() != nullptr);`.
What I wanted is to make sure we do have a back-edge block before traversing the tree.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:612]: Calling mem2reg could solve it (as long as the other steps here stop generating phi nodes). But right now, this pass doesn't work with the legacy pass manager, so I need to add a wrapper. Since this impacts code outside of this backend, I wanted to look into this later.
And since I don't have the pass, I have phi nodes to handle, to made the pass generate them for exits.

I think it would be nice not to have those phi-node constraints, so yes, that's the plan for the next steps: remove all phi-nodes, so I can avoid creating those branches.

About the optimization: do you mean by the driver? Because clang does not optimize them away.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:1165]: Mmm looks like this was indeed redundant since the simplifyBranch is done before. Removed this code, all tests are still passing.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:631]: That's way better, updated the code

File: llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp:878]: Yes, if we add other instruction taking the address, we need to update this code (one of the reasons I needed to update it was that I added new instructions taking BB addresses).

If we don't remove them, the BB are marked as "address taken". The generic ASM printer will then emit empty basic blocks at the end of the function, with a comment saying this empty BB is emitted because its address was taken by another BB.
I saw a PR to address this as the AsmPrinter level (since we never need this AFAIK), but seems like it's being discussed. So in the meantime, I need to clean-up those BlockAddr.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:507]: Yes, that's correct. Turns out this function is not used anymore as getting the continue construct when the loop is not well-structurized is not possible that way (as you said, no guarantees the continue doesn't branch to something else than the header.
Cleaned-up

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:514]: Updated the comment, it returns both the loop construct and the continue construct. Those can only be split later on.
Removed the unused parameter.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:783]: Ah right, optimizations. Rephrased the comment then to capture this possibility, and explain what we must do to emit legal SPIR-V.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp]: Ok turns out the structurizer tests using HLSL files were not running then executing `check-all` :upside_down_face: 
Running those removing this step makes some test break. Adding this again, but changing the comment to fix this bad argument.

File: llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp:878]: @s-perron @Keenuts This is not related to AsmPrinter and emission of labels, rather this change is a part of a series of PR's improving correctness of emitted MIR between passes for branching instructions and thus increasing the number of passing tests when expensive checks are on. This particular is from https://github.com/llvm/llvm-project/pull/106966 and aims to address this error that we have been getting from MachineVerifier:
https://github.com/llvm/llvm-project/blob/300161761df54f5f85630a8ad0e170d09d119ee3/llvm/lib/CodeGen/MachineVerifier.cpp#L697

Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp:878]: Ah it's the machine verifier, not the AsmPrinter! thanks.
But yes, it's the labels being generated at the end of the blob with "address taken, but removed by codegen" this additional code is addressing.

**Comment by VyacheslavLevytskyy** [comment]
I didn't dive deeply into the subject, but I'd be eager to assist in a somewhat indirect manner. I've tried the PR locally to check this with SYCL and machine verifier (expensive checks: on). The former luckily didn't notice any changes, but the latter gives some insights. Probably I see one potential issue to address with respect to the validity of code between passes, namely after SPIRV pre legalizer.

37 of 63 test cases in the 'structurizer' folder are failing when running with 'llc -verify-machineinstrs', and it looks like the repeat offender is "Virtual register defs don't dominate all uses", like in

```
  ...
  %45:iid(s32) = ASSIGN_TYPE %59:iid(s32), %56:type(s64)
  %95:iid(s32) = nsw G_ADD %96:iid, %97:iid
  %96:iid(s32) = GET_ID %45:iid(s32)
  %97:iid(s32) = GET_ID %26:iid(s32)
  %47:iid(s32) = nsw ASSIGN_TYPE %95:iid(s32), %56:type(s64)
  G_STORE %47:iid(s32), %5:pid(p0) :: (store (s32) into %ir.i)
  G_BR %bb.9
  ...

*** Bad machine code: Virtual register defs don't dominate all uses. ***
- function:    main
- v. register: %97
```

No existing tests with '-verify-machineinstrs' start to fail, so I guess there are no problems with code modifications wrt. the general/existing branching logic, but rather with the new use case introduced by the PR/added instructions.

Another weird thing is that llvm specifically build with expensive checks on is now hangs (or maybe hits a very long timeout, I haven't yet a patience to check) while running our test suite, so I'm not quite sure if this PR introduce new issues for Machine Verifier in general, in the part of existing test cases where we haven't yet inserted '-verify-machineinstrs'.

File: llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp:878]: I guess it doesn't matter much and anyway I'm not 100% sure so please never mind. It's just that, as far as I remember, we addressed your report https://github.com/llvm/llvm-project/pull/87823#issuecomment-2113002300 (resulted from changes in https://github.com/llvm/llvm-project/pull/87823) in the later PR https://github.com/llvm/llvm-project/pull/92390 and it's not about setAddressTakenIRBlock() but rather about https://github.com/llvm/llvm-project/blob/df2a2a072a010d757c23237afc9b7c721702fe2b/llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp#L880

**Comment by VyacheslavLevytskyy** [comment]
To be absolutely sure I've just cleanly rebuilt llvm with `-DLLVM_ENABLE_EXPENSIVE_CHECKS=ON` and I can confirm that our test suite is hanging/hitting a timeout now. It's slightly more worrying than 'llc -verify-machineinstrs' fails.

**Comment by VyacheslavLevytskyy** [comment]
The culprit is `test/CodeGen/SPIRV/structurizer/cf.do.continue.ll`.

**Comment by Keenuts** [comment]
Ok, so I've figured out 3 issues:

1. A bug in my sorting of merge instructions. If LHS == RHS, I returned LHS < RHS, that was wrong. Executed outside of GDB, it was very slow to show the backtrace and crash, so tests looked like it was hanging, but it was just crashing.
2. An existing bug in the SPIRVPreLegalizer: the GET_ID instruction insertion point is wrong. Seems like I found a quick fix.
3. When SPIRVGlobalRegistry is used to create OpTypeBool, it creates it not in the main BB, but in the current location. Meaning if you then want this type on a BB located before the last insertion point, you'll end-up with a "use before declare".

Fixing those 3 allowed me to get later in the verification process. I now have 2 more types of issues:
- OpBranch/OpBranchConditional are not considered to be valid basic block terminators
- OpLoopMerge are taking BB operands, but verifier expected register operand.

I'd be tempted to fix the issue 1 (the actual iterator issue), but leave the rest unfixed in this PR (which is already huge).
Then, in distrinct PRs, fir 1 and 2, and then address the next ones.

**Comment by VyacheslavLevytskyy** [comment]
That's really great, thank you.

> OpBranch/OpBranchConditional are not considered to be valid basic block terminators
In lib/Target/SPIRV/SPIRVInstrInfo.td lines 625-632 we actually set isTerminator=1 for both of those, but let's double check then.

> ... leave the rest unfixed in this PR (which is already huge)
Sure!

**Comment by VyacheslavLevytskyy** [comment]
> OpLoopMerge are taking BB operands, but verifier expected register operand.

Hopefully, the fix may be as simple as to change lib/Target/SPIRV/SPIRVInstrInfo.td line 620-621 `ID:$merge, ID:$continue` into `unknown:$merge, unknown:$continue` as in line 626 for OpBranch:
`def OpBranch: Op<249, (outs), (ins unknown:$label), "OpBranch $label">;`

**Comment by Keenuts** [comment]
> > OpLoopMerge are taking BB operands, but verifier expected register operand.
> 
> Hopefully, the fix may be as simple as to change lib/Target/SPIRV/SPIRVInstrInfo.td line 620-621 `ID:$merge, ID:$continue` into `unknown:$merge, unknown:$continue` as in line 626 for OpBranch: `def OpBranch: Op<249, (outs), (ins unknown:$label), "OpBranch $label">;`

Oh thanks!
Turns out the barrier think can also be solved with just an `isBarrier = 1` in this td file.

**Comment by Keenuts** [comment]
Rebased on main.
Also had to apply a few changes:
 - fixed valid expensive checks issues caused by this PR (irreflexible iterator)
This was because the BB order check was saying `A < A`.
 - fixed codegen instability caused by the block sorting.
 
 To satisfy both the SPIR-V spec and the MIR requirements, I had to be stricter on the block-ordering. The initial sorting I had was incomplete, and I had to use the partial topological sorting. But this sorting had a slight issue: when 2 BB had the same order, their order was not defined (same rank). I added the traversal index in addition to consistently compare those blocks.

@VyacheslavLevytskyy There is one test I disabled: `instructions/ret-type.ll`.
The failure can be avoided by not calling `sortBlocks` in `SPIRVPrepareFunction.cpp`.
I checked the IR, and it seems the only change is that the true label is sorted before the false label. And seems like the pointer type deduction fails in that case. Let me know if that's OK with you.

**Comment by VyacheslavLevytskyy** [comment]
Thank you @Keenuts! I'll check it today against my test suites and get back to you. I'd like to to check what's wrong with llvm/test/CodeGen/SPIRV/instructions/ret-type.ll very briefly.

Meanwhile one more question, just out of curiosity. Between the pass of SPIRVPrepareFunctions.cpp with  its new `sortBlocks(F)` and up to SPIRVEmitIntrinsicsPass, just before IRTranslator, we should see a lot of optimization going on when we run with -O1,-O2,-O3. I'm curious how `sortBlocks(F)` and other changes would behave when inserted before vs. after those optimizations applied? In other words, I wonder what is the best place to do this sort of transformation in SPIRV Backend passes and does it depend on with/without optimization choice.

**Comment by Keenuts** [comment]
> I'll check it today against my test suites and get back to you.

Thanks!


> I'm curious how `sortBlocks(F)` and other changes would behave when inserted before vs. after those optimizations applied?

I'm not sure either. I'd be tempted to say we should put the `sortBlock` after passes which move blocks around (like the structurizer), but many others should be fine (DCE could remove blocks/edges, but shouldn't change the order no?)

**Comment by VyacheslavLevytskyy** [comment]
All looks good!

After this is merged I plan to have a look into:
* this PR introduces new tests which fails with "extensive checks on", but they are not related to the PR's logic -- it's an existing issue with correct place to insert SPIRV type definitions required by assign type instructions, and
* llvm/test/CodeGen/SPIRV/instructions/ret-type.ll fails due to reordering of return statements -- I see the reason, but it has no local and simple fix, so probably I should spend some time soon to improve this.

**Review by VyacheslavLevytskyy — APPROVED** [review]

**Comment by Keenuts** [comment]
I created https://github.com/llvm/llvm-project/issues/108845 and https://github.com/llvm/llvm-project/issues/108844 for some of the extensive checks isues and added those to my schedule.
Thanks for the review!

**Comment by Keenuts** [comment]
@michalpaszkowski let me know when I can merge this (I see no warning on my local build, neither on the CI)

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (michalpaszkowski) [inline llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:616]: After loop unrolling or unswitching, we may have loops with `SwitchInst` (instead of `BranchInst`) in the loop header.

**Comment by michalpaszkowski** [comment]
@Keenuts Apologies for the delay! The patch looks good to me. I added a couple of minor comments (above), but if the pass would be extended in subsequent commits, please feel free to address these in the next pull request or reply later. Thank you for working on this, this is really exciting work!

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:954]: right, fixed!

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:1128]: It would? If SplitInstruction is the first, `isMergeInstruction` would return false, meaning we'd call `splitBasicBlockBefore` on this instruction, which means creating an empty BB with just a branch to this instruction no?

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:616]: That's a good point. And to be honest I'm not entirely sure right now.
I added an assert, explaining I made this assumption, but it might be something to revisit.
In addition to that, switches are a bit of a specific case I need to iterate on. Some scenarios are not completely covered.

**Comment by Keenuts** [comment]
Thanks! No worries!
Applied the few changes. And yes, there will be iterations on this. For example, merge-exit pass is not completely finished, and other specific scenarios. But this will be done through smaller PRs.
I also need to look into how to test this since people want me to revert the spirv-sim utility.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (michalpaszkowski) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:1128]: Yes, you are right, started overthinking this a bit. Marking as resolved.

File: llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp
Comment (michalpaszkowski) [reply llvm/lib/Target/SPIRV/SPIRVStructurizer.cpp:616]: Yes, that's okay, thank you for clarifying!

**Review by michalpaszkowski — APPROVED** [review]

**Comment by Keenuts** [comment]
rebased. Will merge tomorrow morning if CI is green.

**Pull request merged**
